# Analysis of next-gen sequencing data
*July 2019* 

This folder contains scripts and guides for general analysis and figure generation of next-gen sequencing data which has been previously performed in the johnstone/kats lab. Many of these scripts are specific for the cluster environment at Peter Mac, and may require alteration. The guides provided are not indeded to be a comprehensive tutorial on analysis, please see the linked resources if you require further help.  If you contribute to this resource, please annotate extensively and define all input and output files within your scripts for ease of use. Please read this [guide](Data_Storage_Guide.md) for more information on how to store your analysed data. 

### Analysis includes: 

**1. RNA-seq**
* 3' RNA-seq (single end) 
* total RNA-seq (paired end) 
* with ERCC spike-in
* Differential gene expression analysis

**2. ChIP-seq**

* Single-end 
* Paired-end
* with DM3 and/or ERCC spike-in
* Traditional peak calling 
* Window-based enrichment analysis with Csaw
* Differential peak calling
* PolII ChIP-seq analysis

**3. ATAC-seq**
* Single -end 
* Paired end 

**4. Reduced Representation Bisulfite Sequencing**

* Single end 
* Paired end with enhanced mapping efficiency
* Differential methylation analysis

**5. M6A-seq**
* Paired-end

**6. CRISPR screen analysis**
