# ATAC-seq Methods

*For publication purposes* 

Example from [Bcor loss perturbs myeloid differentiation and promotes leukaemogenesis](https://www.nature.com/articles/s41467-019-09250-6) 

ATAC-seq was performed using a previously published protocol69. In brief, 5x104 sorted KL cells were lysed and nuclei washed, followed by chromatin tagmentation using TD1 enzyme and TDE buffer supplied in Illumina Nextera DNA prep kit (cat#FC-121-1030). Chromatin was tagmented at 37°C for 30 minutes, then purified using the Minielute kit (Qiagen). A DNA library was prepared for sequencing by ligating adapters and amplifying using KAPA hot start polymerase (KAPA biosystems) according to manufacturer’s instructions then purifying using Minielute kit (Qiagen). QC was performed using D1000 high-sensitivity screen tape with 4200 TapeStation Instrument (Agilent Technologies) before size selection for fragments between 150bp and 700bp using a Pippin Prep System (Sage Science). Libraries were pooled and sequenced with 75bp single end sequencing to a depth of 15-20 x 106 reads per sample on a NextSeq (Illumina). bcl2fastq (v2.17.1.14) was used for de-multiplexing. The Fastq files generated by sequencing were aligned to the mouse reference genome (GRCm38/mm10) using bowtie2 (v2.3.3). Samtools (v1.4.1) was used for manipulation of SAM and BAM files. Data was visualised with deeptools (v2.5.3) alongside ChIP-seq data.
